src/Stratis.Bitcoin/Base/InvalidBlockHashStore.cs

```diff
@@ -0,0 +1,154 @@
+﻿using System;
+using System.Collections.Generic;
+using NBitcoin;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Base
+{
+    /// <summary>
+    /// Contract of a store of block header hashes that are considered invalid.
+    /// </summary>
+    public interface IInvalidBlockHashStore
+    {
+        /// <summary>
+        /// Check if a block is marked as invalid.
+        /// </summary>
+        /// <param name="hashBlock">The block hash to check.</param>
+        /// <returns><c>true</c> if the block is marked as invalid, <c>false</c> otherwise.</returns>
+        bool IsInvalid(uint256 hashBlock);
+
+        /// <summary>
+        /// Marks a block as invalid.
+        /// </summary>
+        /// <param name="hashBlock">The block hash to mark as invalid.</param>
+        /// <param name="rejectedUntil">Time in UTC after which the block is no longer considered as invalid, or <c>null</c> if the block is to be considered invalid forever.</param>
+        void MarkInvalid(uint256 hashBlock, DateTime? rejectedUntil = null);
+    }
+
+    /// <summary>
+    /// In memory store of invalid block header hashes.
+    /// </summary>
+    /// <remarks>
+    /// The store has a limited capacity. When a new block header hash is marked as invalid 
+    /// once the capacity is reached, the oldest entry is removed and no longer considered invalid.
+    /// <para>
+    /// Entries with specified expiration time are either removed just like other entries - i.e. during 
+    /// an add operation when a capacity is reached, or they are removed when they are touched 
+    /// and it is detected that their expiration time is no longer in the future.
+    /// </para>
+    /// </remarks>
+    public class InvalidBlockHashStore : IInvalidBlockHashStore
+    {
+        /// <summary>Default value for the maximal number of hashes we can store.</summary>
+        public const int DefaultCapacity = 1000;
+
+        /// <summary>A provider of the date and time.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        /// <summary>Lock object to protect access to <see cref="invalidBlockHashes"/> and <see cref="orderedHashList"/>.</summary>
+        private readonly object lockObject = new object();
+
+        /// <summary>
+        /// Collection of blocks that are to be considered invalid. If the value of the entry is not <c>null</c>, 
+        /// the entry is considered invalid only for a certain amount of time.
+        /// </summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<uint256, DateTime?> invalidBlockHashes;
```

Comment:
Isn't it better to make this a ConcurrentDictionary? Thus removing the need for a lock?

Reply:
no, there are two structures that are synchronized, so we need a lock

Reply:
Yes but only in the ```MarkInvalid``` method. The ```IsInvalid``` method does not require a lock if you use a Concurrent dictionary. Its more appropriate to use ConcurrentDictionary if you can do so 👍

Reply:
i disagree

CD has its own lock, so what you are saying is that you want 2 locks instead of 1

but most importantly, the logic in MarkInvalid is nontrivial and i can't see how you can achieve it easily with CD 

what i'm doing ther is that i'm updating an entry if the existing entry has some specific properties, this is something CD does not support thread-safely

i can't examine CD item, make a test on that value and then change it in one atomic step with just internal CD lock

Reply:
OK yeah, leave as is, I understand now what you are trying to do @Aprogiena 👍

Reply:
@fassadlr I tend to prefer the Concurrent ns merely for simplicity of the code, but apro is more of a `lock` guys both are fine.